staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/downloader.go

```diff
@@ -21,32 +21,62 @@ import (
 	"fmt"
 	"net/http"
 	"strings"
+	"sync"
 
 	"k8s.io/apiserver/pkg/authentication/user"
 	"k8s.io/apiserver/pkg/endpoints/request"
-	"k8s.io/kube-openapi/pkg/cached"
 	"k8s.io/kube-openapi/pkg/validation/spec"
 )
 
+type CacheableDownloader interface {
+	UpdateHandler(http.Handler)
+	Get() (*spec.Swagger, string, error)
+}
+
 // cacheableDownloader is a downloader that will always return the data
 // and the etag.
 type cacheableDownloader struct {
+	name       string
 	downloader *Downloader
 	handler    http.Handler
 	etag       string
 	spec       *spec.Swagger
+
+	// mutex protects the handler which can be replaced
+	mutex sync.Mutex
 }
 
 // NewCacheableDownloader creates a downloader that also returns the etag, making it useful to use as a cached dependency.
-func NewCacheableDownloader(downloader *Downloader, handler http.Handler) cached.Value[*spec.Swagger] {
+func NewCacheableDownloader(apiServiceName string, downloader *Downloader, handler http.Handler) CacheableDownloader {
 	return &cacheableDownloader{
+		name:       apiServiceName,
 		downloader: downloader,
 		handler:    handler,
 	}
 }
+func (d *cacheableDownloader) UpdateHandler(handler http.Handler) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+	d.handler = handler
+}
+
+func (d *cacheableDownloader) getHandler() http.Handler {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+	return d.handler
+}
 
 func (d *cacheableDownloader) Get() (*spec.Swagger, string, error) {
-	swagger, etag, status, err := d.downloader.Download(d.handler, d.etag)
+	handler := d.getHandler()
+	spec, etag, err := d.getWithHandler(handler)
+	if err != nil {
+		return spec, etag, fmt.Errorf("failed to download %v: %v", d.name, err)
+	}
+	return spec, etag, err
+}
+
+func (d *cacheableDownloader) getWithHandler(h http.Handler) (*spec.Swagger, string, error) {
```

Comment:
is there a `get` without handler? The name sounds like there would be. But wait, I think I see where this is coming from. You shouldn't have to receive the handler here.

Reply:
I've updated with the atomic pointer, but kept this function. Separating the two functions lets us decorate the error with the apiservice name. lmk if you have thoughts, thanks!

Reply:
Sure, I still think it's weird to pass to `getWithHandler` one of it's own member though.

Reply:
Maybe just do ...
```golang
spec, etag, err := d.get()
...
func (d *cacheableDownloader) get() (*spec.Swagger, string, error) {
    h := *d.handler.Load()
...
```

Reply:
Updated